server/src/ankaios_server/server_state.rs

```diff
@@ -0,0 +1,843 @@
+                &"requestId".into(),
+                request_complete_state.request_id.to_owned().into(),
+            )?;
+
+            for field in &request_complete_state.field_mask {
+                if let Some(value) = current_complete_state.get(&field.into()) {
+                    return_state.set(&field.into(), value.to_owned())?;
+                } else {
+                    log::debug!(
+                        concat!(
+                        "Result for CompleteState incomplete, as requested field does not exist:\n",
+                        "   request_id: {:?}\n",
+                        "   field: {}"),
+                        request_complete_state.request_id,
+                        field
+                    );
+                    continue;
+                };
+            }
+
+            return_state.try_into().map_err(|err: serde_yaml::Error| {
+                format!("The result for CompleteState is invalid: '{}'", err)
+            })
+        } else {
+            Ok(current_complete_state)
+        }
+    }
+
+    pub fn get_workloads_for_agent(&self, agent_name: &String) -> Vec<WorkloadSpec> {
+        self.state
+            .current_state
+            .workloads
+            .clone()
+            .into_values()
+            // [impl->swdd~agent-from-agent-field~1]
+            .filter(|workload_spec| workload_spec.agent.eq(agent_name))
+            .collect()
+    }
+
+    pub fn update(
+        &mut self,
+        new_state: CompleteState,
+        update_mask: Vec<String>,
+    ) -> Result<AddedDeletedWorkloads, UpdateStateError> {
+        // [impl->swdd~server-state-rejects-state-with-cyclic-dependencies~1]
+        match update_state(&self.state, new_state, update_mask) {
+            Ok(new_state) => {
+                let cmd = extract_added_and_deleted_workloads(
+                    &self.state.current_state,
+                    &new_state.current_state,
+                );
+
+                if let Some((added_workloads, deleted_workloads)) = cmd {
+                    let start_nodes: Vec<&String> = added_workloads
+                        .iter()
+                        .filter_map(|w| {
+                            if !w.dependencies.is_empty() {
+                                Some(&w.name)
+                            } else {
+                                None
+                            }
+                        })
+                        .collect();
+
+                    log::debug!(
+                        "Execute cyclic dependency check with start_nodes = {:?}",
+                        start_nodes
+                    );
+
+                    if let Some(workload_part_of_cycle) =
+                        cycle_check::dfs(&new_state.current_state, Some(start_nodes))
+                    {
+                        return Err(UpdateStateError::CycleInDependencies(
+                            workload_part_of_cycle,
+                        ));
+                    }
```

Comment:
We should trace at the place where the check is done.
```suggestion
                    // [impl->swdd~server-state-rejects-state-with-cyclic-dependencies~1]
                    log::debug!(
                        "Execute cyclic dependency check with start_nodes = {:?}",
                        start_nodes
                    );
                    if let Some(workload_part_of_cycle) =
                        cycle_check::dfs(&new_state.current_state, Some(start_nodes))
                    {
                        return Err(UpdateStateError::CycleInDependencies(
                            workload_part_of_cycle,
                        ));
                    }
```